src/renderers/webgl/WebGLRenderStates.js

```diff
@@ -56,7 +56,7 @@ function WebGLRenderState() {
 
 function WebGLRenderStates() {
 
-	let renderStates = new WeakMap();
+	let renderStates = new Map();
```

Comment:
Assuming the `dispose` method is correctly called where it needs to be called, then we don't need a `WeakMap` here. With `WeakMap`, we can not iterate on it in order to do the below cleanup.

Reply:
Turning `renderStates` to a `Map` prevents garbing collection at other places. I have to reconsider the implications of this change...

Reply:
When an instance of `Scene` is not used anymore, the entry in the `renderStates` should not prevent the scene from being garbage collected. When the scene is garbage collected, its event listeners should also be removed and thus there should be no references to render states anymore. Hence, I'm not sure it's necessary to explicitly remove the event listener.

Are you able to demonstrate a potential memory leak with a live exampel?

Reply:
The reconsideration is simple: any code that handles an instance of `WebGLRenderStates` simply needs to call `dispose` when it is finished using the instance.

Another issue is that if someone is using the `WebGLRenderStates` class for any other reason (whatever it may be) they also need a way to explicitly signal that they are done with the instance, _without having to lose references to everything that the instance holds_, which is currently the case with the `WeakMap` because the only way for things to get be explicitly cleaned up is if the user explicitly loses all references to objects that the user assumes are held within. In my opinion, and because there already is a dispose method, making the `dispose` method the ultimate signal of explicit disposal is the best approach, otherwise disposal is difficult.

Plus with WeakMaps, we can't explicitly dispose of the event handlers. Losing references is the only way to explicitly do that (which forces a pattern whereby the user must always create `new` instances rather than re-using them).

Maybe for this particular class it isn't a big deal however. But the above are my thoughts on disposal in general (if something has an explicit API to clean things up, there's no need for WeakMaps and we may as well make sure all cleanup is explicit).